tools/asset-sync/assets.ps1

```diff
@@ -0,0 +1,521 @@
+Set-StrictMode -Version 3
+
+# See https://stackoverflow.com/a/14440066 for a succinct explanation. We will be using
+# this methodology for any global variables that can be considered constants
+$REPO_ROOT = Resolve-Path (Join-Path $PSScriptRoot ".." "..")
+$ASSETS_STORE = (Join-Path $REPO_ROOT ".assets")
+
+. (Join-Path $REPO_ROOT "eng" "common" "scripts" "common.ps1")
+
+<#
+.SYNOPSIS
+Checks the contents of a directory, then returns a tuple of booleans @("assetsJsonPresent", "isRootFolder").
+
+.DESCRIPTION
+Evaluates a directory by checking its contents. First value of the tuple is whether or not a "assets.json" file 
+is present A "root" directory is one where a assets.json is present OR where we are as far up the file tree as 
+we can possibly ascend. 
+
+.PARAMETER TargetPath
+The path we will evaulate to detect either a root directory or a .git folder.
+#>
+Function Evaluate-Target-Dir {
+    param (
+        [Parameter(Mandatory = $true)]
+        [string] $TargetPath
+    )
+
+    $isFile = Test-Path $TargetPath -PathType Leaf
+    $foundConfig = $false
+    $foundRoot = $false
+
+    if ($isFile) {
+        throw "Evaluated a file `"$TargetPath`" as a directory. Exiting."
+    }
+
+    # we need to force to ensure we grab hidden files in the directory dump
+    $files = Get-ChildItem -Force $TargetPath
+
+    foreach($file in $files)
+    {
+        if($file.Name.ToLower() -eq "assets.json"){
+            $foundConfig = $true
+        }
+
+        # need to ensure that if we're outside a git repo, we will still break out eventually.
+        if(($file.Name.ToLower() -eq ".git") -or -not (Split-Path $TargetPath)){
+            $foundRoot = $true
+        }
+    }
+
+    return @(
+        $foundConfig, $foundRoot
+    )
+}
+
+<#
+.SYNOPSIS
+Gets the relative path of the assets json within the target repo.
+
+.PARAMETER Config
+The config
+#>
+Function AscendToRepoRoot {
```

Comment:
Repo root can be found without crawling:  `git -C $pathForManipulation rev-parse --show-toplevel`
Path root can also be found without crawl: `[IO.Path]::GetPathRoot($pathForManipulation)`
If we're not using $foundConfig, maybe one of those is better.

Reply:
I think I can definitely use this. Thanks for the suggestion!

Reply:
Damnit. I can't.

I ended up with 
```pwsh
<#
.SYNOPSIS
Given a target path, evaluate the root of the containing git repo.

.PARAMETER StartPath
Our start point. If this path does not exist within a git repo, an error will be thrown.
#>
Function GetRepoRoot {
    param(
        [Parameter(Mandatory = $false)]
        [string] $StartPath
    )

    $pathForManipulation = $StartPath

    if (!(Test-Path $pathForManipulation)) {
        throw "The given path $($StartPath) does not exist."
    }

    # the rev-parse methodology works given a folder name, but _not_ given a specific file
    if (Test-Path $pathForManipulation -PathType Leaf) {
        $pathForManipulation, $fileName = Split-Path $pathForManipulation
    }

    $result = (git -C $pathForManipulation rev-parse --show-toplevel)

    if ($LASTEXITCODE -gt 0){
        throw "Unable to evaluate the repo root for path `"$($pathForManipulation)`""
    }

    return $result
}
```

But due to how we initially check out the repo (filter: 0, no blobs existing yet) the show top level fails with

> fatal: invalid gitfile format: C:/repo/sdk-tools/tools/asset-sync/.testruns/f4fba398-eedf-47fa-9a97-ebad8192a8de/.git

Because we haven't yet initialized. I will be cleaning up the `Evaluate-Git-Dir` and the usage in this function, but I think it needs to stick around.